enhancements/machine-config/manage-boot-images.md

```diff
@@ -0,0 +1,427 @@
+
+Here are some YAML examples that describes operators in each of these modes:
+##### Enabled
+```
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+metadata:
+  name: default
+  labels:
+spec:
+  managedBootImageConfig:
+    mode: Enabled
+```
+##### Disabled
+```
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+metadata:
+  name: default
+  labels:
+spec:
+  managedBootImageConfig:
+    mode: Disabled
+```
+##### MatchSelector
+```
+apiVersion: operator.openshift.io/v1
+kind: MachineConfiguration
+metadata:
+  name: default
+  labels:
+spec:
+  managedBootImageConfig:
+    mode: CustomConfig
+    CustomConfig:
+      machineSetSelector:
+        matchLabels:
+          machineconfiguration.openshift.io/mco-managed-machineset: ""        
+```
+Note: While in this mode, the label added to the selector will have to be added to the `machineset` object.
+
+A [ValidatingAdmissionPolicy](https://kubernetes.io/docs/reference/access-authn-authz/validating-admission-policy/) will be implemented via an MCO manifest that will restrict updating the `ManagedBootImageConfig` object to only supported platforms(initially, just GCP). This will be updated as we phase in support for other platforms. Here is a sample policy that would do this:
+
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "managed-bootimages-platform-check"
+spec:
+  failurePolicy: Fail
+  paramKind:
+    apiVersion: config.openshift.io/v1
+    kind: Infrastructure
+  matchConstraints:
+    resourceRules:
+    - apiGroups:   ["operator"]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["MachineConfiguration"]
+  validations:
+    - expression: "has(object.spec.MachineBootImageConfig) && param.status.platformStatus.Type != `GCP`"
+      message: "This feature is only supported on these platforms: GCP"
+```
+This would need an accompanying binding:
+```
+apiVersion: admissionregistration.k8s.io/v1beta1
+kind: ValidatingAdmissionPolicyBinding
+metadata:
+  name: "managed-bootimages-platform-check-binding"
+spec:
+  policyName: "managed-bootimages-platform-check"
+  validationActions: [Deny]
+  paramRef:
+    name: "cluster"
+    namespace: "default"
+```
+#### Tracking boot image history
+
+This proposal will also introduce a new CR, `MachineSetBootImageHistory` for tracking boot image history. As a starting point, here is a stub type definition for this:
+
+```
+type MachineSetBootImageHistory struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   MachineSetBootImageHistorySpec   `json:"spec,omitempty"`
+	Status MachineSetBootImageHistoryStatus `json:"status,omitempty"`
+}
+
+// MachineSetBootImageHistorySpec defines the desired state of MachineSetBootImageHistory
+type MachineSetBootImageHistorySpec struct {
+	MachineSetName string                   `json:"machineSetName"`
```

Comment:
Rather than a name, you may want to use a reference type that allows multiple fields. If this API will work for both Machine API and Cluster API, we probably want to include the API group, name and namespace in this reference at the minimum.

If this object exists in the same namespace as the MachineSet, you can probably drop namespace above

Reply:
Makes sense, thanks!